### PR TITLE
[pcl] Move VTK to an optional feature

### PR DESCRIPTION
**Pull request by here-mikelley**
**Describe the pull request**

- What does your PR fix? Fixes issue #

This doesn't fix an issue.

- Which triplets are supported/not supported? Have you updated the CI baseline?

All triplets should be supported. Making VTK optional also allows pcl to be compiled as a dynamic library under Linux.

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?

Yes

### CODE REVIEW

File: ports/pcl/CONTROL [hunk 1]
```diff
@@ -2,7 +2,7 @@ Source: pcl
 Version: 1.9.1-11
```
Comment (NancyLi1013) [inline ports/pcl/CONTROL]: The version should be updated as `1.9.1-12`.

File: ports/pcl/portfile.cmake [hunk 2]
```diff
@@ -44,7 +45,6 @@ vcpkg_configure_cmake(
         -DWITH_LIBUSB=OFF
         -DWITH_PNG=ON
         -DWITH_QHULL=ON
-        -DWITH_VTK=ON
```
Comment (NancyLi1013) [inline ports/pcl/portfile.cmake]: This option is not needed to add here. Since it is included in `${FEATURE_OPTIONS}`.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

File: ports/pcl/CONTROL [hunk 3]
```diff
@@ -2,7 +2,7 @@ Source: pcl
...
 Homepage: https://github.com/PointCloudLibrary/pcl
 Description: Point Cloud Library (PCL) is open source library for 2D/3D image and point cloud processing.
-Build-Depends: eigen3, flann, qhull, vtk, libpng, boost-system, boost-filesystem, boost-thread, boost-date-time, boost-iostreams, boost-random, boost-foreach, boost-dynamic-bitset, boost-property-map, boost-graph, boost-multi-array, boost-signals2, boost-ptr-container, boost-uuid, boost-interprocess, boost-asio
+Build-Depends: eigen3, flann, qhull, libpng, boost-system, boost-filesystem, boost-thread, boost-date-time, boost-iostreams, boost-random, boost-foreach, boost-dynamic-bitset, boost-property-map, boost-graph, boost-multi-array, boost-signals2, boost-ptr-container, boost-uuid, boost-interprocess, boost-asio
```
Comment (JackBoosY) [inline ports/pcl/CONTROL]: Since vtk is a required dependency before, other ports may depend on this feature. Should we add `Default-Feature: vtk` here?

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/pcl/portfile.cmake (hunk 2 above)
Comment (here-mikelley) [reply ports/pcl/portfile.cmake]: I think you're misreading this. I removed the ```-DWITH_VTK`` in my commit.

File: ports/pcl/CONTROL (hunk 3 above)
Comment (here-mikelley) [reply ports/pcl/CONTROL]: Sounds good. I added it as default parameter in an updated commit.

File: ports/pcl/portfile.cmake (hunk 2 above)
Comment (NancyLi1013) [reply ports/pcl/portfile.cmake]: Sorry for my mistake.

**Review by NancyLi1013 — APPROVED** [review]

**Comment by JackBoosY** [comment]
@here-mikelley Does this issue ready to review?

File: ports/pcl/portfile.cmake (hunk 2 above)
Comment (kelleyma49) [reply ports/pcl/portfile.cmake]: No need to apologize - it's an easy mistake to make!

**Comment by kelleyma49** [comment]
@JackBoosY I believe it is. Please review.

**Comment by here-mikelley** [comment]
@JackBoosY I believe it is. Please review.

**Review by JackBoosY — APPROVED** [review]

**Comment by dan-shaw** [comment]
@here-mikelley Why does VTK need to be an optional feature? You can install `pcl` without `VTK` using `vcpkg install pcl[core]`

**Comment by here-mikelley** [comment]
> @here-mikelley Why does VTK need to be an optional feature? You can install `pcl` without `VTK` using `vcpkg install pcl[core]`

I must be misunderstanding `Build-Depends` and feature lists. Isn't `Build-Depends` a list of required dependencies? I thought that the feature list disabled default features and enabled optional features. I interpret `pcl[core]` meaning that all build dependencies specified in `Build-Depends` are required and built.

**Comment by here-mikelley** [comment]
@dan-shaw , I just tried your suggestion and it still installs vtk:
```
PS /Users/mikelley/vcpkg> ./vcpkg/vcpkg install pcl[core]:x64-osx
Computing installation plan...
The following packages will be built and installed:
  * boost-asio[core]:x64-osx
  * boost-coroutine[core]:x64-osx
  * boost-signals2[core]:x64-osx
  * flann[core]:x64-osx
  * openssl[core]:x64-osx
    pcl[core]:x64-osx
  * vtk[core]:x64-osx
Additional packages (*) will be modified to complete this operation.
```

**Comment by NancyLi1013** [comment]
@here-mikelley 
Could you please resolve the conflicts?

**Comment by here-mikelley** [comment]
> @here-mikelley
> Could you please resolve the conflicts?

It looks as though VTK was forced to be OFF in the latest code on master. My commit moved VTK to a feature that can be turned off/on, so I'm not sure how to resolve this conflict. We should revisit this change once VTK is enabled again.

File: ports/pcl/CONTROL [hunk 4]
```diff
@@ -1,8 +1,9 @@
 Source: pcl
-Version: 1.9.1-11
+Version: 1.9.1-12
 Homepage: https://github.com/PointCloudLibrary/pcl
 Description: Point Cloud Library (PCL) is open source library for 2D/3D image and point cloud processing.
-Build-Depends: eigen3, flann, qhull, vtk, libpng, boost-system, boost-filesystem, boost-thread, boost-date-time, boost-iostreams, boost-random, boost-foreach, boost-dynamic-bitset, boost-property-map, boost-graph, boost-multi-array, boost-signals2, boost-ptr-container, boost-uuid, boost-interprocess, boost-asio
+Build-Depends: eigen3, flann, qhull, libpng, boost-system, boost-filesystem, boost-thread, boost-date-time, boost-iostreams, boost-random, boost-foreach, boost-dynamic-bitset, boost-property-map, boost-graph, boost-multi-array, boost-signals2, boost-ptr-container, boost-uuid, boost-interprocess, boost-asio
+Default-Features: vtk
```
Comment (JackBoosY) [inline ports/pcl/CONTROL:6]: I think we should not set vtk as pcl default feature.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/pcl/CONTROL (hunk 4 above)
Comment (here-mikelley) [reply ports/pcl/CONTROL:6]: I left it as a default feature as that was essentially the previous behavior. I'm fine with removing it. 

Note that, as mentioned in my previous comment, VTK is turned OFF on master, so my changes won't work until VTK is turned back on.

**Comment by NancyLi1013** [comment]
Hi @here-mikelley 

 Feature `vtk `build failed on x86-windows:

`F:\10449\vcpkg\buildtrees\pcl\src\pcl-1.9.1-f2ec88a858\io\src\vtk_lib_io.cpp(369): error C2664: 'int vtkCellArray::GetNextCell(vtkIdList *)': cannot convert argument 2 from 'vtkIdType *' to 'const vtkIdType *&'`

Could you please try to look into it?

**Comment by NancyLi1013** [comment]
Hi @here-mikelley 
Is work still being done for this PR?

**Comment by here-mikelley** [comment]
VTK is currently disabled for pcl (see https://github.com/microsoft/vcpkg/blob/7960648b4d922888e7a5c2a4c3ae9e153eab0374/ports/pcl/portfile.cmake#L49.) I've paused work on this until it's enabled again.

**Comment by NancyLi1013** [comment]
Sorry for the late reply.
Seems you have removed this line in this PR.

`        -DWITH_VTK=OFF # disabled due to API changes in 9.0`

**Comment by BillyONeal** [comment]
Manually tested:

* `pcl[core]:x86-windows` (also checked that it did not install vtk)
* `pcl[core,vtk]:x86-windows`
* `pcl[core]:x86-windows-static`
* `pcl[core,vtk]:x86-windows-static`
* `pcl[core]:x64-osx` (also checked that it did not install vtk)
* `pcl[core,vtk]:x64-osx`

**Review by BillyONeal — APPROVED** [review]

**Review by JackBoosY — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks for your contribution!

**Pull request merged**